### PR TITLE
Improve abort/cancellation support for shard transfer

### PR DESCRIPTION
**Pull request by ffuugoo**
Tracked in #2432.

This PR continues the work in #2918 and makes most of the shard transfer methods and functions cancel safe.

There are a bunch of `TODO`s added by this PR, but I think they should be left as-is for now, and can be worked on in the following PRs.

### All Submissions:

* [x] Contributions should target the `dev` branch. Did you create your branch from `dev`?
* [x] Have you followed the guidelines in our Contributing document?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../../pulls) for the same update/change?

### New Feature Submissions:

1. [x] Does your submission pass tests?
2. [x] Have you formatted your code locally using `cargo +nightly fmt --all` command prior to submission?
3. [x] Have you checked your code using `cargo clippy --all --all-features` command?

### Changes to Core Features:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [ ] Have you written new tests for your core changes, as applicable?
* [ ] Have you successfully ran tests with your changes locally?

### CODE REVIEW

File: lib/collection/src/shards/transfer/shard_transfer.rs
Comment (timvisee) [inline lib/collection/src/shards/transfer/shard_transfer.rs:302]: Maybe we do want some function that returns a `TempFile` to prevent this.

File: lib/collection/src/shards/transfer/shard_transfer.rs
Comment (timvisee) [inline lib/collection/src/shards/transfer/shard_transfer.rs:802]: We want to handle it as regular error. More specifically, we want to call `revert_queue_proxy_local()`. We probably call this later as well as part of aborting the snapshot, but doing it here as well is a good thing to be sure.

File: lib/collection/src/shards/transfer/shard_transfer.rs
Comment (timvisee) [inline lib/collection/src/shards/transfer/shard_transfer.rs:823]: I think this branch is fine; doing nothing on cancel.

File: lib/collection/src/shards/queue_proxy_shard.rs
Comment (timvisee) [inline lib/collection/src/shards/queue_proxy_shard.rs:57]: I don't like that we expose the `max_ack_version` here. I get that it's to make the function non-async.

But I don't think we should expose this to the outside.

Can we find some middle ground here?

File: lib/collection/src/shards/replica_set/mod.rs
Comment (timvisee) [inline lib/collection/src/shards/replica_set/mod.rs:384]: Probably a good idea. It currently relies on the timeout to eventually drop the thread.

Moving the cancellation flag in and checking on each loop is probably already good enough.

File: lib/collection/src/shards/replica_set/shard_transfer.rs
Comment (timvisee) [inline lib/collection/src/shards/replica_set/shard_transfer.rs:320]: Yes.

File: lib/collection/src/shards/transfer/shard_transfer.rs
Comment (timvisee) [reply lib/collection/src/shards/transfer/shard_transfer.rs:802]: What about changing this `match` to:

```rust
match &result {
    Ok(()) => break,

    Err(err) => {
        // Revert queue proxy if we still have any to clean up
        if let Some(shard) = shards_holder.read().await.get_shard(&transfer.shard_id) {
            shard.revert_queue_proxy_local().await;
        }

        if let CollectionError::Cancelled { .. } = err {
            break;
        }

        log::error!(
            "Failed to transfer shard {} -> {}: {err}",
            transfer.shard_id,
            transfer.to,
        );
    },
}
```

Comment (ffuugoo) [reply lib/collection/src/shards/transfer/shard_transfer.rs:799]: Ooooh, that's a nice suggestion, I like it. 👌

Comment (timvisee) [reply lib/collection/src/shards/transfer/shard_transfer.rs:799]: It does reorder logging after reverting the queue proxy, but I guess it's fine.

Comment (timvisee) [reply lib/collection/src/shards/transfer/shard_transfer.rs:799]: Now that I think of it, we can probably remove the whole match and do an 'early return' if `Ok(())`` to clean this up even further.

File: lib/collection/src/shards/queue_proxy_shard.rs
Comment (ffuugoo) [reply lib/collection/src/shards/queue_proxy_shard.rs:57]: I don't like it either, but I don't see other way to do it.

And it's either make `QueueProxyShard::now` sync, or `ShardReplicaSet::queue_proxify_local` is *not* cancel safe and there are cancellation tokens everywhere. :/

**Comment by timvisee** [comment]
Very nice work! Happy to approve once we resolve the above comments.

File: lib/collection/src/shards/replica_set/mod.rs
Comment (ffuugoo) [reply lib/collection/src/shards/replica_set/mod.rs:384]: It's just rather non-trivial to do (because we have a sync task blocked on a cond-var), so I left it as `TODO` for now.

File: lib/collection/src/shards/transfer/shard_transfer.rs
Comment (ffuugoo) [inline lib/collection/src/shards/transfer/shard_transfer.rs]: I don't think this is resolved, cause there's `replica_set.remove_local().await` below that may not be cancel safe, and so `handle_transferred_shard_proxy` may not be cancel safe.

File: lib/collection/src/shards/queue_proxy_shard.rs
Comment (timvisee) [reply lib/collection/src/shards/queue_proxy_shard.rs:57]: Yeah this is tricky.

Does using explicit cancellation tokens in `queue_proxify_local` look that bad though? Because to be honest, I don't like all the restructuring we're required to do just for this any better.

File: lib/collection/src/shards/transfer/shard_transfer.rs
Comment (timvisee) [reply lib/collection/src/shards/transfer/shard_transfer.rs]: Fair enough, but the same TODO is still there a few lines below.

Comment (timvisee) [reply lib/collection/src/shards/transfer/shard_transfer.rs]: My eye now catches the `add_remote` line which may be troublesome as well. 

Alright, lets put it back. Sorry.

File: lib/collection/src/shards/queue_proxy_shard.rs
Comment (ffuugoo) [reply lib/collection/src/shards/queue_proxy_shard.rs:57]: Hmmm... Maybe you're right. We'd have to propagate cancellation token:

- from the shard transfer task to `transfer_shard`
- from `transfer_shard` to `transfer_snapshot`
- and refactor `transfer_snapshot` so that it would `cancel_on_token` everything before and after `queue_proxify_local` and propagate into it

Which might be reasonable.

File: lib/collection/src/shards/transfer/shard_transfer.rs
Comment (ffuugoo) [reply lib/collection/src/shards/transfer/shard_transfer.rs]: > Fair enough, but the same TODO is still there a few lines below.

My bad. It's how I format these things. I put two of these `TODO`s, one at the top, and one at a particular thing that need work (if there's a specific item identified).

File: lib/collection/src/shards/queue_proxy_shard.rs
Comment (timvisee) [reply lib/collection/src/shards/queue_proxy_shard.rs:57]: I guess we should keep it as-is for now, since changing this again may be a lot of work. We can revisit this idea later. The current implementation seems logically correct to me. :+1:

File: lib/collection/src/shards/transfer/shard_transfer.rs
Comment (ffuugoo) [reply lib/collection/src/shards/transfer/shard_transfer.rs:301]: Another refactoring of `create_shard_snapshot` (or adding a separate method) 🥲

File: lib/collection/src/shards/transfer/shard_transfer.rs
Comment (ffuugoo) [reply lib/collection/src/shards/transfer/shard_transfer.rs]: I've restructured `spawn_transfer_task`, what do you think? 🤔

File: lib/collection/src/shards/transfer/shard_transfer.rs
Comment (ffuugoo) [reply lib/collection/src/shards/transfer/shard_transfer.rs]: Removed.

File: lib/collection/src/shards/transfer/shard_transfer.rs
Comment (ffuugoo) [inline lib/collection/src/shards/transfer/shard_transfer.rs:809]: This may be more readable, though...

```suggestion
            if matches!(result, Ok(Ok(())) | Err(_)) {
```

Comment (timvisee) [reply lib/collection/src/shards/transfer/shard_transfer.rs:809]: Yeah lets do this one. I was a bit confused by the negation.

File: lib/collection/src/shards/transfer/shard_transfer.rs
Comment (timvisee) [inline lib/collection/src/shards/transfer/shard_transfer.rs:811]: What about this? We use this pattern in some other places as well.

```suggestion
            let is_cancelled = result.is_err();
            let is_ok = matches!(result, Ok(Ok(())));
            let is_err = matches!(result, Ok(Err(_)));

            // If `transfer_shard` returned error or task was cancelled...
            if is_cancelled || is_err {
                // Revert queue proxy if we still have any to prepare for the next attempt
                if let Some(shard) = shards_holder.read().await.get_shard(&transfer.shard_id) {
                    shard.revert_queue_proxy_local().await;
                }
            }

            // If `transfer_shard` returned success or task was cancelled...
            if is_ok || is_cancelled {
                break;
            }
```

Comment (ffuugoo) [reply lib/collection/src/shards/transfer/shard_transfer.rs:811]: That's... much better, actually. I don't know why I didn't implement it like that in the first place. 🤦‍♀️

Comment (ffuugoo) [reply lib/collection/src/shards/transfer/shard_transfer.rs]: Done.

Comment (timvisee) [reply lib/collection/src/shards/transfer/shard_transfer.rs]: I also always get blind for improvements like this after refactoring some code a million times. :laughing: 

> Done.

Awesome!

**Comment by timvisee** [comment]
Let's merge once :green_circle:

**Review by timvisee — APPROVED** [review]

**Pull request merged**
